File: citizenapp/android/app/src/main/kotlin/com/crcfrcn/citizenapp/SquareMediaChannel.kt
package com.crcfrcn.citizenapp

import android.content.Context
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/** 公民广场媒体原生入口；通道字段统一使用 snake_case。 */
class SquareMediaChannel(
    messenger: BinaryMessenger,
    context: Context,
) {
    private val transcoder = SquareVideoTranscoder(context.applicationContext)
    private val channel = MethodChannel(messenger, CHANNEL_NAME)

    init {
        channel.setMethodCallHandler(::handle)
    }

    fun dispose() {
        transcoder.cancel()
        channel.setMethodCallHandler(null)
    }

    private fun handle(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "capabilities" -> result.success(transcoder.capabilities())
            "transcode_video" -> {
                try {
                    transcoder.transcode(SquareVideoRequest.from(call), result)
                } catch (error: Exception) {
                    result.error("square_media_invalid", error.message, null)
                }
            }
            "cancel_video" -> {
                transcoder.cancel()
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    companion object {
        const val CHANNEL_NAME = "citizenapp/square_media"
    }
}

data class SquareVideoRequest(
    val inputPath: String,
    val outputPath: String,
    val coverPath: String,
    val maxWidth: Int,
    val maxHeight: Int,
    val videoBitrate: Int,
    val totalPeakBitrate: Int,
    val audioBitrate: Int,
    val audioSampleRate: Int,
    val maxFrameRate: Int,
    val keyFrameIntervalSeconds: Int,
    val maxDurationSeconds: Int,
    val maxBytes: Long,
    val coverMaxEdge: Int,
    val coverQuality: Int,
    val coverMaxBytes: Int,
) {
    init {
        require(inputPath.isNotBlank() && outputPath.isNotBlank() && coverPath.isNotBlank()) {
            "媒体路径不能为空"
        }
        require(maxWidth in 2..1920 && maxHeight in 2..1080) { "视频尺寸不合法" }
        require(videoBitrate > 0 && audioBitrate > 0 && totalPeakBitrate > 0) { "视频码率不合法" }
        require(audioSampleRate == 48_000) { "音频采样率必须为 48kHz" }
        require(maxFrameRate in 1..30 && keyFrameIntervalSeconds == 2) { "帧率或关键帧间隔不合法" }
        require(maxDurationSeconds in 1..10_800 && maxBytes in 1..3_000_000_000L) { "视频上限不合法" }
        require(coverMaxEdge in 2..720 && coverQuality in 1..100 && coverMaxBytes in 1..512_000) {
            "视频封面规则不合法"
        }
    }

    companion object {
        fun from(call: MethodCall): SquareVideoRequest = SquareVideoRequest(
            inputPath = call.requiredString("input_path"),
            outputPath = call.requiredString("output_path"),
            coverPath = call.requiredString("cover_path"),
            maxWidth = call.requiredInt("max_width"),
            maxHeight = call.requiredInt("max_height"),
            videoBitrate = call.requiredInt("video_bitrate"),
            totalPeakBitrate = call.requiredInt("total_peak_bitrate"),
            audioBitrate = call.requiredInt("audio_bitrate"),
            audioSampleRate = call.requiredInt("audio_sample_rate"),
            maxFrameRate = call.requiredInt("max_frame_rate"),
            keyFrameIntervalSeconds = call.requiredInt("key_frame_interval_seconds"),
            maxDurationSeconds = call.requiredInt("max_duration_seconds"),
            maxBytes = call.requiredLong("max_bytes"),
            coverMaxEdge = call.requiredInt("cover_max_edge"),
            coverQuality = call.requiredInt("cover_quality"),
            coverMaxBytes = call.requiredInt("cover_max_bytes"),
        )
    }
}

private fun MethodCall.requiredString(key: String): String =
    argument<String>(key)?.takeIf { it.isNotBlank() }
        ?: throw IllegalArgumentException("$key 缺失")

private fun MethodCall.requiredInt(key: String): Int =
    (argument<Number>(key)?.toInt()) ?: throw IllegalArgumentException("$key 缺失")

private fun MethodCall.requiredLong(key: String): Long =
    (argument<Number>(key)?.toLong()) ?: throw IllegalArgumentException("$key 缺失")

File: citizenapp/ios/Runner/SquareMediaChannel.swift
import Flutter
import Foundation

/// 公民广场媒体原生通道；跨端字段统一使用 snake_case。
final class SquareMediaChannel {
  static let channelName = "citizenapp/square_media"

  private let channel: FlutterMethodChannel
  private let transcoder = SquareVideoTranscoder()

  init(binaryMessenger: FlutterBinaryMessenger) {
    channel = FlutterMethodChannel(
      name: Self.channelName,
      binaryMessenger: binaryMessenger
    )
    channel.setMethodCallHandler { [weak self] call, result in
      self?.handle(call, result: result)
    }
  }

  deinit {
    transcoder.cancel()
    channel.setMethodCallHandler(nil)
  }

  private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "capabilities":
      result(transcoder.capabilities())
    case "transcode_video":
      do {
        let request = try SquareVideoRequest(arguments: call.arguments)
        transcoder.transcode(request) { transcodeResult in
          DispatchQueue.main.async {
            switch transcodeResult {
            case let .success(value):
              result(value)
            case let .failure(error):
              result(
                FlutterError(
                  code: "square_media_transcode_failed",
                  message: error.localizedDescription,
                  details: nil
                )
              )
            }
          }
        }
      } catch {
        result(
          FlutterError(
            code: "square_media_invalid",
            message: error.localizedDescription,
            details: nil
          )
        )
      }
    case "cancel_video":
      transcoder.cancel()
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}

struct SquareVideoRequest {
  let inputPath: String
  let outputPath: String
  let coverPath: String
  let maxWidth: Int
  let maxHeight: Int
  let videoBitrate: Int
  let totalPeakBitrate: Int
  let audioBitrate: Int
  let audioSampleRate: Int
  let maxFrameRate: Int
  let keyFrameIntervalSeconds: Int
  let maxDurationSeconds: Int
  let maxBytes: Int64
  let coverMaxEdge: Int
  let coverQuality: Int
  let coverMaxBytes: Int

  init(arguments: Any?) throws {
    guard let values = arguments as? [String: Any] else {
      throw SquareMediaError.invalidRequest("视频参数不是对象")
    }
    inputPath = try values.requiredString("input_path")
    outputPath = try values.requiredString("output_path")
    coverPath = try values.requiredString("cover_path")
    maxWidth = try values.requiredInt("max_width")
    maxHeight = try values.requiredInt("max_height")
    videoBitrate = try values.requiredInt("video_bitrate")
    totalPeakBitrate = try values.requiredInt("total_peak_bitrate")
    audioBitrate = try values.requiredInt("audio_bitrate")
    audioSampleRate = try values.requiredInt("audio_sample_rate")
    maxFrameRate = try values.requiredInt("max_frame_rate")
    keyFrameIntervalSeconds = try values.requiredInt("key_frame_interval_seconds")
    maxDurationSeconds = try values.requiredInt("max_duration_seconds")
    maxBytes = try values.requiredInt64("max_bytes")
    coverMaxEdge = try values.requiredInt("cover_max_edge")
    coverQuality = try values.requiredInt("cover_quality")
    coverMaxBytes = try values.requiredInt("cover_max_bytes")

    guard maxWidth >= 2, maxWidth <= 1920,
          maxHeight >= 2, maxHeight <= 1080,
          videoBitrate > 0, audioBitrate > 0, totalPeakBitrate > 0,
          audioSampleRate == 48_000,
          maxFrameRate > 0, maxFrameRate <= 30,
          keyFrameIntervalSeconds == 2,
          maxDurationSeconds > 0, maxDurationSeconds <= 10_800,
          maxBytes > 0, maxBytes <= 3_000_000_000,
          coverMaxEdge > 0, coverMaxEdge <= 720,
          coverQuality > 0, coverQuality <= 100,
          coverMaxBytes > 0, coverMaxBytes <= 512_000
    else {
      throw SquareMediaError.invalidRequest("视频参数超出允许范围")
    }
  }
}

enum SquareMediaError: LocalizedError {
  case invalidRequest(String)
  case unsupported(String)
  case processing(String)
  case cancelled

  var errorDescription: String? {
    switch self {
    case let .invalidRequest(message), let .unsupported(message), let .processing(message):
      return message
    case .cancelled:
      return "视频处理已取消"
    }
  }
}

private extension Dictionary where Key == String, Value == Any {
  func requiredString(_ key: String) throws -> String {
    guard let value = self[key] as? String, !value.isEmpty else {
      throw SquareMediaError.invalidRequest("\(key) 缺失")
    }
    return value
  }

  func requiredInt(_ key: String) throws -> Int {
    guard let number = self[key] as? NSNumber else {
      throw SquareMediaError.invalidRequest("\(key) 缺失")
    }
    return number.intValue
  }

  func requiredInt64(_ key: String) throws -> Int64 {
    guard let number = self[key] as? NSNumber else {
      throw SquareMediaError.invalidRequest("\(key) 缺失")
    }
    return number.int64Value
  }
}
